File: c/sorting_algorithms/tree_sort.c
#include <stdio.h>
#include <stdlib.h>

// Definicion del nodo del arbol binario
struct Node {
    int data;
    struct Node* left;
    struct Node* right;
};

// Crear un nuevo nodo
struct Node* newNode(int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = data;
    node->left = node->right = NULL;
    return node;
}

// Insertar un nodo en el arbol
struct Node* insert(struct Node* node, int data) {
    if (node == NULL)
        return newNode(data);
    if (data < node->data)
        node->left = insert(node->left, data);
    else if (data > node->data)
        node->right = insert(node->right, data);
    return node;
}

// Recorrido in-order del arbol
void inorder(struct Node* root, int arr[], int* index) {
    if (root != NULL) {
        inorder(root->left, arr, index);
        arr[(*index)++] = root->data;
        inorder(root->right, arr, index);
    }
}

// Tree Sort
void treeSort(int arr[], int n) {
    struct Node* root = NULL;

    // Insertar elementos en el arbol
    for (int i = 0; i < n; i++) {
        root = insert(root, arr[i]);
    }

    // Recorrido in-order para ordenar
    int index = 0;
    inorder(root, arr, &index);
}

// Imprimir un array
void printArray(int arr[], int size) {
    for (int i = 0; i < size; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

int main() {
    int arr[] = {5, 3, 7, 1, 9, 4, 6};
    int n = sizeof(arr) / sizeof(arr[0]);

    printf("\nArray original:\n");
    printArray(arr, n);

    treeSort(arr, n);

    printf("\nArray ordenado:\n");
    printArray(arr, n);

    return 0;
}

File: python/sorting_algorithms/tree_sort.py
class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

# Insertar un nodo en el arbol
def insert(root, key):
    if root is None:
        return Node(key)
    if key < root.key:
        root.left = insert(root.left, key)
    elif key > root.key:
        root.right = insert(root.right, key)
    return root

# Recorrido in-order del arbol
def inorder(root, sorted_list):
    if root is not None:
        inorder(root.left, sorted_list)
        sorted_list.append(root.key)
        inorder(root.right, sorted_list)

# Tree Sort
def tree_sort(arr):
    if not arr:
        return []

    root = None
    for key in arr:
        root = insert(root, key)

    sorted_list = []
    inorder(root, sorted_list)
    return sorted_list

# Ejemplo de uso
arr = [5, 3, 7, 1, 9, 4, 6]
print(f"\nArray original: {arr}")
sorted_arr = tree_sort(arr)
print(f"\nArray ordenado: {sorted_arr}")